xarray/util/generate_reductions.py

```diff
@@ -3,294 +3,398 @@
 
-        >>> {calculation}(){maybe_skipna}{maybe_mincount}"""
 
+class GenericReductionGenerator(ReductionGenerator):
+    def generate_code(self, method):
+        extra_kwargs = [kwarg.call for kwarg in method.extra_kwargs if kwarg.call]
 
-def generate_method(
-    obj: str,
-    docref: str,
-    method: str,
-    skipna: bool,
-    example_generator: Callable,
-    array_method: Optional[str] = None,
-):
-    if not array_method:
-        array_method = method
+        if self.datastructure.numeric_only:
+            extra_kwargs.append(f"numeric_only={method.numeric_only},")
 
-    if obj == "Dataset":
-        if method in NUMERIC_ONLY_METHODS:
-            numeric_only_call = "\n            numeric_only=True,"
+        if extra_kwargs:
+            extra_kwargs = textwrap.indent("\n" + "\n".join(extra_kwargs), 12 * " ")
         else:
-            numeric_only_call = "\n            numeric_only=False,"
-    else:
-        numeric_only_call = ""
-
-    kwarg = collections.namedtuple("kwarg", "docs kwarg call")
-    if skipna:
-        skip_na = kwarg(
-            docs=textwrap.indent(_SKIPNA_DOCSTRING, "        "),
-            kwarg="\n        skipna: bool = True,",
-            call="\n            skipna=skipna,",
-        )
-    else:
-        skip_na = kwarg(docs="", kwarg="", call="")
-
-    if method in MIN_COUNT_METHODS:
-        min_count = kwarg(
-            docs=textwrap.indent(_MINCOUNT_DOCSTRING, "        "),
-            kwarg="\n        min_count: Optional[int] = None,",
-            call="\n            min_count=min_count,",
-        )
-    else:
-        min_count = kwarg(docs="", kwarg="", call="")
-
-    return TEMPLATE_REDUCTION.format(
-        obj=obj,
-        docref=docref,
-        method=method,
-        array_method=array_method,
-        extra_dim="""If ``None``, will reduce over all dimensions
-            present in the grouped variable.""",
-        extra_args="",
-        skip_na=skip_na,
-        min_count=min_count,
-        numeric_only_call=numeric_only_call,
-        example=example_generator(obj=obj, method=method),
-    )
-
-
-def render(obj: str, cls: str, docref: str, example_generator: Callable):
-    yield CLASS_PREAMBLE.format(obj=obj, cls=cls)
-    yield generate_method(
-        obj,
-        method="count",
-        docref=docref,
-        skipna=False,
-        example_generator=example_generator,
-    )
-    for method in BOOL_REDUCE_METHODS:
-        yield generate_method(
-            obj,
-            method=method,
-            docref=docref,
-            skipna=False,
-            array_method=f"array_{method}",
-            example_generator=example_generator,
-        )
-    for method in NAN_REDUCE_METHODS:
-        yield generate_method(
-            obj,
-            method=method,
-            docref=docref,
-            skipna=True,
-            example_generator=example_generator,
-        )
+            extra_kwargs = ""
+        return f"""        return self.reduce(
```

Comment:
Using backslash to escape a newline might make this a little more readable?
```suggestion
        return f"""\
        return self.reduce(
```